zfs-snapshots.py

```diff
@@ -0,0 +1,100 @@
+#!/usr/bin/env python3
+import os
+import subprocess
+from functools import reduce, partial
+from itertools import groupby
+from operator import itemgetter, add
+from prometheus_client import CollectorRegistry, Gauge, generate_latest
+
+
+def row_to_metric(metric: Gauge, row):
+    return metric.labels(pool=row[0][0], volume=row[0][1]).set(row[1])
+
+
+def collect_metrics(metric: Gauge, it):
+    return list(map(partial(row_to_metric, metric), it))
+
+
+def zfs_parse_line(line):
+    cols = line.split("\t")
+    rest, snapshot = cols[0].rsplit("@", 1)
+    pool = rest
+    volume = None
+    if "/" in rest:
+        pool, volume = rest.split("/", 1)
+        volume = "/" + volume
+    return [pool, volume, snapshot, *map(int, cols[1:])]
+
+
+def zfs_list_snapshots():
+    cmd = [
+        "zfs",
+        "list",
+        "-p",
+        "-H",
+        "-t",
+        "snapshot",
+        "-o",
+        "name,used,creation",
+    ]
+    # zfs list can be relatively slow (couple of seconds)
+    # Use Popen to incrementally read from stdout to not waste further time
+    popen = subprocess.Popen(
+        cmd, stdout=subprocess.PIPE, env=dict(os.environ, LC_ALL="C")
+    )
+    for stdout_line in iter(popen.stdout.readline, ""):
+        stdout_line = stdout_line.strip()
+        if stdout_line == b"":
+            break
+        yield stdout_line.decode("utf-8")
+    return_code = popen.wait()
+    if return_code:
+        raise subprocess.CalledProcessError(return_code, cmd)
+
+
+def aggregate_rows(rows, index, operator):
+    return map(
+        lambda row: (row[0], reduce(operator, map(itemgetter(index), row[1]), 0)), rows
+    )
+
+
+NAMESPACE = "zfs_snapshot"
+LABEL_NAMES = ["pool", "volume"]
+
+
+def main():
+    registry = CollectorRegistry()
+    latest_snapshot_taken_metric = Gauge(
+        "latest_snapshot_taken",
+        "Timestamp of the latest snapshot",
+        labelnames=LABEL_NAMES,
+        namespace=NAMESPACE,
+        registry=registry,
+    )
+    space_usage_metric = Gauge(
+        "space_usage",
```

Comment:
nit: `foo_used` is more widely accepted naming convention
```suggestion
        "space_used",
```

Reply:
Done